### ISSUE TITLE
Request for inference code / model definition (bsrnn_legacy) for zero-shot academic benchmark

### ISSUE THREAD

**Issue by Panda-Lorrain** [TaurenMountain/PS4#1]
Hi @TaurenMountain,

First, congratulations on the PS4 paper (arXiv:2607.08111) and ranking **2nd overall** in the REAL-T Challenge (best SIM and F1). The proxy-supervised joint training design — ASR cross-entropy + speaker-similarity + frame-level VAD + perceptual quality — is, in our view, exactly the right direction for bridging the perception-recognition gap in TSE, and we'd like to cite it as a reference architecture.

We are researching **target-speaker ASR** (recognizing a target speaker's voice commands under noise and multi-speaker overlap, given a short ~1.8s enrollment) and would like to use your released PS4 checkpoint (`TaurenMountain/PS4`, `checkpoint_epoch037.pt`) for a **zero-shot academic benchmark** on our internal evaluation set, comparing against our own pipeline (diarization + target-timeline extraction + Qwen3-ASR).

However, the HuggingFace release contains only the checkpoint **without the model definition class (`bsrnn_legacy.py`) or an inference script**, and the public `REAL-TSE/wesep-real-tse` repo does not include this class (`get_model("BSRNN")` exits with code 1; only `TSE_BSRNN_SPK` is registered). The state_dict key naming (`separator.separation` / `spk_model.layer*` / `mask.*` / `BN.*` / `spk_encoder` / `preEmphasis`) does not match the public `TSE_BSRNN_SPK` keys (`sep_model.*` / `spk_ft.*`), so we are unable to load the weights for inference.

Would it be possible to share:
1. The model definition class (`bsrnn_legacy.py`) and/or a minimal inference script + config, or
2. Point us to where these can be obtained?

We commit to:
- **Research-only, non-commercial** use
- **Proper citation** of your paper (Ning et al., 2026) in any resulting report
- **Not redistributing** the code or checkpoint

Thank you very much for your time, and for releasing PS4 and REAL-PS4 to the community!

Best,
Panda_Lorrain

**Comment by ningwanyi** [comment]
Hi @Panda-Lorrain ,

Thanks for your interest in PS4! Sorry about the confusion — the original release only had the checkpoint weights and the training code, but the model definition file (`bsrnn_legacy.py`) was missing from the public repository, which is why you couldn't load the checkpoint.

I've now updated both repositories:

1. HuggingFace Model ([TaurenMountain/PS4](https://huggingface.co/TaurenMountain/PS4)):
Added a self-contained inference.py script. Works with just torch, torchaudio, and numpy — no other dependencies needed.

```
pip install torch torchaudio numpy
python inference.py \
    --checkpoint checkpoint_epoch037.pt \
    --mix mix.wav \
    --enroll target.wav \
    --output result.wav
```

2. GitHub Training Code ([TaurenMountain/PS4](https://github.com/TaurenMountain/PS4)):

- Added `inference.py` (same as HuggingFace)
- Added `wesep_ps4/` — a minimal dependency package containing the legacy BSRNN model, ECAPA-TDNN speaker encoder, and all - necessary modules. No need to clone or modify the upstream wesep repository anymore.
- Updated `train.py` and `run_train.sh` to use `wesep_ps4/` automatically.

You can now clone the repo and run training directly. Let me know if you run into any other issues!
